---
title: "Finemapping of genetic regions in inbred mice"
output: BiocStyle::html_document
author: "Matthias Munz"
vignette: >
    %\VignetteIndexEntry{Finemapping}
    %\VignetteEngine{knitr::rmarkdown}
    %\VignetteEncoding{UTF-8}
---


```{r, include = FALSE}
knitr::opts_chunk$set(
    collapse = TRUE,
    comment = "#>"
)
```

# Introduction
This **R** package provides methods for **genetic finemapping** in **inbred mice** 
by taking advantage of their **very high homozygosity rate** (>95%). 

For one ore more chromosomal regions (**GRCm38**), method `finemap` extracts homozygous SNVs for which the allele differs between two sets of strains (e.g. case vs controls) and outputs respective causal SNV/gene candidates.

# Installation
```{r setup, eval=FALSE}
if(!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("MouseFM")
```

# Loading package
```{r}
library(MouseFM)
```

# Example function calls
Available mouse strains
```{r}
avail_strains()
```

Call finemap to finemap a specific region
```{r}
res = finemap(chr="chr7",
              strain1=c("C57BL_6J","C57L_J","CBA_J","NZB_B1NJ"),
              strain2=c("C3H_HEJ","MOLF_EiJ","NZW_LacJ","WSB_EiJ","SPRET_EiJ"),
              impact=c("HIGH", "MODERATE", "LOW"))

res[1:10,]
```

View meta information
```{r}
comment(res)
```

Annotate with consequences (Not recommended for large queries!)
```{r eval=FALSE}
cons = annotate_consequences(res, "mouse")
```

Annotate with genes
```{r eval=FALSE}
genes = annotate_mouse_genes(res, flanking = 50000)
```

# Output of Session Info
The output of `sessionInfo()` on the system
on which this document was compiled:
```{r}
sessionInfo()
```
