# You can learn more about package authoring with RStudio at:
#
#   http://r-pkgs.had.co.nz/
#
# Some useful keyboard shortcuts for package authoring:
#
#   Install Package:           'Cmd + Shift + B'
#   Check Package:             'Cmd + Shift + E'
#   Test Package:              'Cmd + Shift + T'
#   roxygen2::roxygenise()


#'@description Available strains
#'@return Vector
#'@export
avail_strains = function(){

  df = data.frame(id = c("129P2_OlaHsd",
                         "129S1_SvImJ",
                         "129S5SvEvBrd",
                         "A_J",
                         "AKR_J",
                         "BALB_cJ",
                         "BTBR_Tplus_Itpr3tf_J",
                         "BUB_BnJ",
                         "C3H_HeH",
                         "C3H_HeJ",
                         "C57BL_10J",
                         "C57BL_6J",
                         "C57BL_6NJ",
                         "C57BR_cdJ",
                         "C57L_J",
                         "C58_J",
                         "CAST_EiJ",
                         "CBA_J",
                         "DBA_1J",
                         "DBA_2J",
                         "FVB_NJ",
                         "I_LnJ",
                         "KK_HiJ",
                         "LEWES_EiJ",
                         "LP_J",
                         "MOLF_EiJ",
                         "NOD_ShiLtJ",
                         "NZB_B1NJ",
                         "NZO_HlLtJ",
                         "NZW_LacJ",
                         "PWK_PhJ",
                         "RF_J",
                         "SEA_GnJ",
                         "SPRET_EiJ",
                         "ST_bJ",
                         "WSB_EiJ",
                         "ZALENDE_Ei"),
                  strain = c("129P2/OlaHsd",
                             "129S1/SvImJ",
                             "129S5/SvEvBrd",
                             "A/J",
                             "AKR/J",
                             "BALB/cJ",
                             "BTBR T + Itpr3tf/J",
                             "BUB/BnJ",
                             "C3H/HeH",
                             "C3H/HeJ",
                             "C57BL/10J",
                             "C57BL/6",
                             "C57BL/6NJ",
                             "C57BR/cdJ",
                             "C57L/J",
                             "C58/J",
                             "CAST/EiJ",
                             "CBA/J",
                             "DBA/1J",
                             "DBA/2J",
                             "FVB/NJ",
                             "I/LnJ",
                             "KK/HiJ",
                             "LEWES/EiJ",
                             "LP/J",
                             "MOLF/EiJ",
                             "NOD/ShiLtJ",
                             "NZB/B1NJ",
                             "NZO/HlLtJ",
                             "NZW/LacJ",
                             "PWK/PhJ",
                             "RF/J",
                             "SEA/GnJ",
                             "SPRET/EiJ",
                             "ST/bJ",
                             "WSB/EiJ",
                             "ZALENDE/EiJ"))

  return(df)
}


#'@description Available consequence and impact types
#'@return Data frame
#'@export
avail_consequences = function(){
  df = data.frame(consequence = c("splice_acceptor_variant",
                                   "splice_donor_variant",
                                   "stop_gained",
                                   "stop_lost",
                                   "splice_region_variant",
                                   "incomplete_terminal_codon_variant",
                                   "stop_retained_variant",
                                   "synonymous_variant",
                                   "missense_variant",
                                   "coding_sequence_variant",
                                   "mature_miRNA_variant",
                                   "5_prime_UTR_variant",
                                   "3_prime_UTR_variant",
                                   "non_coding_transcript_exon_variant",
                                   "intron_variant",
                                   "NMD_transcript_variant",
                                   "non_coding_transcript_variant",
                                   "upstream_gene_variant",
                                   "downstream_gene_variant",
                                   "intergenic_variant",
                                   "initiator_codon_variant"),
                  impact = c("HIGH",
                             "HIGH",
                             "HIGH",
                             "HIGH",
                             "LOW",
                             "LOW",
                             "LOW",
                             "LOW",
                             "MODERATE",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER",
                             "MODIFIER"),
                    so_description = c("A splice variant that changes the 2 base region at the 3' end of an intron",
                                     "A splice variant that changes the 2 base region at the 5' end of an intron",
                                     "A sequence variant whereby at least one base of a codon is changed, resulting in a premature stop codon, leading to a shortened transcript",
                                     "A sequence variant where at least one base of the terminator codon (stop) is changed, resulting in an elongated transcript",
                                     "A sequence variant in which a change has occurred within the region of the splice site, either within 1-3 bases of the exon or 3-8 bases of the intron",
                                     "A sequence variant where at least one base of the final codon of an incompletely annotated transcript is changed",
                                     "A sequence variant where at least one base in the terminator codon is changed, but the terminator remains",
                                     "A sequence variant where there is no resulting change to the encoded amino acid",
                                     "A sequence variant, that changes one or more bases, resulting in a different amino acid sequence but where the length is preserved",
                                     "A sequence variant that changes the coding sequence",
                                     "A transcript variant located with the sequence of the mature miRNA",
                                     "A UTR variant of the 5' UTR",
                                     "A UTR variant of the 3' UTR",
                                     "A sequence variant that changes non-coding exon sequence in a non-coding transcript",
                                     "A transcript variant occurring within an intron",
                                     "A variant in a transcript that is the target of NMD",
                                     "A transcript variant of a non coding RNA gene",
                                     "A sequence variant located 5' of a gene",
                                     "A sequence variant located 3' of a gene",
                                     "A sequence variant located in the intergenic region, between genes",
                                     "A codon variant that changes at least one base of the first codon of a transcript."),
                    so_term = c("SO:0001574",
                                "SO:0001575",
                                "SO:0001587",
                                "SO:0001578",
                                "SO:0001630",
                                "SO:0001626",
                                "SO:0001567",
                                "SO:0001819",
                                "SO:0001583",
                                "SO:0001580",
                                "SO:0001620",
                                "SO:0001623",
                                "SO:0001624",
                                "SO:0001792",
                                "SO:0001627",
                                "SO:0001621",
                                "SO:0001619",
                                "SO:0001631",
                                "SO:0001632",
                                "SO:0001628",
                                "SO:0001582"),
                    display_term = c("Splice acceptor variant",
                                     "Splice donor variant",
                                     "Stop gained",
                                     "Stop lost",
                                     "Splice region variant",
                                     "Incomplete terminal codon variant",
                                     "Stop retained variant",
                                     "Synonymous variant",
                                     "Missense variant",
                                     "Coding sequence variant",
                                     "Mature miRNA variant",
                                     "5 prime UTR variant",
                                     "3 prime UTR variant",
                                     "Non coding transcript exon variant",
                                     "Intron variant",
                                     "NMD transcript variant",
                                     "Non coding transcript variant",
                                     "Upstream gene variant",
                                     "Downstream gene variant",
                                     "Intergenic variant",
                                     "Initiator codon variant"))
  return(df)
}


